pkg/node/manager/manager.go

```diff
@@ -340,58 +340,74 @@ func (m *manager) backgroundSyncInterval() time.Duration {
 func (m *manager) backgroundSync(ctx context.Context) error {
-	syncTimer, syncTimerDone := inctimer.New()
-	defer syncTimerDone()
 	for {
 		syncInterval := m.backgroundSyncInterval()
-		log.WithField("syncInterval", syncInterval.String()).Debug("Performing regular background work")
 
-		var errs error
-		// get a copy of the node identities to avoid locking the entire manager
-		// throughout the process of running the datapath validation.
-		nodes := m.GetNodeIdentities()
-		for _, nodeIdentity := range nodes {
-			// Retrieve latest node information in case any event
-			// changed the node since the call to GetNodes()
-			m.mutex.RLock()
-			entry, ok := m.nodes[nodeIdentity]
-			if !ok {
-				m.mutex.RUnlock()
-				continue
-			}
-			m.mutex.RUnlock()
-
-			entry.mutex.Lock()
-			{
-				m.Iter(func(nh datapath.NodeHandler) {
-					if err := nh.NodeValidateImplementation(entry.node); err != nil {
-						log.WithFields(logrus.Fields{
-							"handler": nh.Name(),
-							"node":    entry.node.Name,
-						}).WithError(err).
-							Error("Failed to apply node handler during background sync. Cilium may have degraded functionality. See error message for details.")
-						errs = errors.Join(errs, fmt.Errorf("failed while handling %s on node %s: %w", nh.Name(), entry.node.Name, err))
-					}
-				})
-			}
-			entry.mutex.Unlock()
-
-			m.metrics.DatapathValidations.Inc()
-		}
+		log.WithField("syncInterval", syncInterval.String()).Debug("Starting new iteration of background sync")
+		err := m.singleBackgroundLoop(ctx, syncInterval)
+		log.WithField("syncInterval", syncInterval.String()).Debug("Finished iteration of background sync")
 
 		hr := m.health.NewScope("background-sync")
-		if errs != nil {
-			hr.Degraded("Failed to apply node validation", errs)
+		if err != nil {
+			hr.Degraded("Failed to apply node validation", err)
 		} else {
 			hr.OK("Node validation successful")
 		}
+	}
+}
+
+func (m *manager) singleBackgroundLoop(ctx context.Context, expectedLoopTime time.Duration) error {
+	var errs error
+	// get a copy of the node identities to avoid locking the entire manager
+	// throughout the process of running the datapath validation.
+	nodes := m.GetNodeIdentities()
+	limiter := rate.NewLimiter(
+		rate.Limit(float64(len(nodes))/float64(expectedLoopTime.Seconds())),
+		1, // One token in bucket to amortize for latency of the operation
+	)
+	for _, nodeIdentity := range nodes {
+		if err := limiter.Wait(ctx); err != nil {
+			log.WithError(err).Debug("Error while rate limiting backgroundSync updates")
+		}
 
 		select {
 		case <-ctx.Done():
 			return nil
-		case <-syncTimer.After(syncInterval):
+		default:
+		}
+		// Retrieve latest node information in case any event
+		// changed the node since the call to GetNodes()
+		m.mutex.RLock()
+		entry, ok := m.nodes[nodeIdentity]
+		if !ok {
+			m.mutex.RUnlock()
+			continue
+		}
+		m.mutex.RUnlock()
+
+		// TODO(marseel): Isn't that a bug?
```

Comment:
Imho there is a bug here, I will followup with a separate PR as I didn't want to block this PR on it, as reviewing any lock-related stuff is tricky.

Reply:
Might be worth to create an issue for that and link it here?

Reply:
I will open PR right after we merge this one.